plans/objchange/plan_valid.go

```diff
@@ -0,0 +1,259 @@
+package objchange
+
+import (
+	"fmt"
+
+	"github.com/zclconf/go-cty/cty"
+
+	"github.com/hashicorp/terraform/configs/configschema"
+)
+
+// AssertPlanValid checks checks whether a planned new state returned by a
+// provider's PlanResourceChange method is suitable to achieve a change
+// from priorState to config. It returns a slice with nonzero length if
+// any problems are detected. Because problems here indicate bugs in the
+// provider that generated the plannedState, they are written with provider
+// developers as an audience, rather than end-users.
+//
+// All of the given values must have the same type and must conform to the
+// implied type of the given schema, or this function may panic or produce
+// garbage results.
+//
+// During planning, a provider may only make changes to attributes that are
+// null (unset) in the configuration and are marked as "computed" in the
```

Comment:
Currently this is still an `"or" are marked as computed`. Default values don't change attributes to be computed (and helper/schema won't allow `computed` with a default  itself, though that has slightly different semantics)

Reply:
_This_ function is requiring that _both_ be true... the provider gets to invent a new value if and only if it is unset in config _and_ it is marked as computed.

I realize that this is stricter than what `helper/schema` assumes today, and that is the main reason why we have the legacy SDK opt-out: so that those older, weaker assumptions don't trip up this check with a fatal error. We will still mention the inconsistency in the `WARNING`-level logs as an aid to both us and provider developers in understanding Terraform's emergent behavior if an inconsistency causes a knock-on error in a downstream resource, but we'll allow the SDK to continue its old behavior until we're ready to replace it, and accept that for now the warning logs here will be a little noisy and require human intuition to decide which reported problems are real problems vs. just legacy SDK behaviors.

Moving forward though, any new SDK will need to mark as `Computed` any attribute that has a default -- whether it be a constant default, a default function, or an apply-time default (what the legacy SDK calls "computed": unknown during planning) -- in order for this check to pass. We can then pass that information on to a future function for exporting the provider schemas in a machine-readable way so that users and external software can know whether a particular attribute takes on a default value when not set, or whether it just stays null in that case.

Reply:
Got it, I though this function was allowing the change in first read :+1: